spec/index.adoc

```diff
@@ -560,7 +560,15 @@ In order to call a canister, the user makes a POST request to `/api/v2/canister/
 * `method_name` (`text`): Name of the canister method to call
 * `arg` (`blob`): Argument to pass to the canister method
 
-The HTTP response to this request has an empty body and HTTP status 202, or a HTTP error (4xx or 5xx). Paranoid agents should not trust this response, and use <<http-read-state,`read_state`>> to determine the status of the call.
+The HTTP response to this request can have the following responses:
+
+* 202 HTTP status with empty body. Implying the request was accepted by the IC for further processing. Users should use <<http-read-state,`read_state`>> to determine the status of the call.
+* 200 HTTP status with non-empty body. Implying an execution pre-processing error occurred. The body of the response contains more information about the IC specific error encountered. The body is a CBOR map with the following fields:
```

Comment:
I'm not a fan of HTTP 2xx indicating an error. 2xx generally indicate success - see also [here](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).

Reply:
It was success in the sense the request reached the execution environment. Compared to all alternatives this is maybe the least confusing we found given we want to distinguish between canister/execution errors and IC platform errors.

Reply:
`execution pre-processing error` is this something like failed to decode candid or rejected by guard method? 

Anything above 5xx has no canonical meaning, why not just use 7xx for it? If I as a dev see 200, I am sure everything worked out fine, which it obviously didn't

Reply:
I just saw `All valid status codes are within the range of 100 to 599, inclusive.` in the RFC... blegh.

Looking through the RFC, I feel like for a pre-processing error a lot would be more appropriate than 200. The first sentence of [200](https://www.rfc-editor.org/rfc/rfc9110.html#name-200-ok) is `The 200 (OK) status code indicates that the request has succeeded.`, which is in direct conflict to a pre-processing **error** IMO. [202](https://www.rfc-editor.org/rfc/rfc9110.html#name-202-accepted) feels better, but still it says `accepted for processing, but the processing has not been completed` which is partially true, but processing _has_ completed, with an error. Some codes in the 4xx range seem more valid on first glance (e.g. [406 Not acceptable](https://www.rfc-editor.org/rfc/rfc9110.html#name-406-not-acceptable) or [412 Precondition failed](https://www.rfc-editor.org/rfc/rfc9110.html#name-412-precondition-failed)), but they are talking about headers, which also seems wrong for this case. Maybe [422 Unprocessable content](https://www.rfc-editor.org/rfc/rfc9110.html#name-422-unprocessable-content) would be valid - it talks about `unable to process the contained instructions`.

Reply:
> If I as a dev see 200, I am sure everything worked out fine, which it obviously didn't

Even today, the spec explicitly says that receiving 200 or even 202 means nothing. You should pool the state to make sure your update call made it through. Given the async nature of the protocol I don't think we have much choice here.

Reply:
I agree with Severin that 2xx codes should not be used for platform errors. Aside from developer confusion, it also reduces compatibility with existing (and future) HTTP aware tooling that understands HTTP codes. Differentiating between platform errors and canister errors using the body fields sounds better, in my opinion.

Reply:
There is a difference to be made between the IC/subnet/replica (the application, if you want) and the HTTP API used to access it. The HTTP status codes (whether 2xx, 4xx or 5xx) only refer to the API, which acts as a proxy to the application. To use the envelope and content analogy at the top of the *Call* section, the status codes refer to the delivery of the envelope.

The alternative (and what we have now, for the most part) is to try and make the HTTP API entirely transparent. But that means either:
* Using the same status code to report on both API and IC errors. E.g. using 404 for "no such API endpoint", "no such canister" and "no such canister method". Similarly, and more confusingly, returning 5xx errors both when the API encounters a server error; when there's an internal replica/sandbox error; and when the canister logic produces an error or panics.
* Or designating some made-up 4xx and 5xx status codes to reflect replica or canister errors. At which point (i) they are not immediately understandable to an HTTP client; and (ii) they are not all that different from returning a 2xx status and a body clarifying the error. Furthermore, a 200 or 202 response to a `call` endpoint by no means guarantees that the subnet will ever induct the ingress message into a block. I.e. we already have the case where a replica replies with 200 or 202, only for the ingress message to disappear without a trace.

Maybe part of the confusion also stems from explicitly stating that 202 stands for success and 200 for error. It would probably be cleaner and clearer to say that a 2xx response means that the API delivered the envelope to the replica and the outcome, if any, is in the body of the response.

Reply:
> Even today, the spec explicitly says that receiving 200 or even 202 means nothing. [...] Given the async nature of the protocol I don't think we have much choice here.

Fair enough. I still don't like how 2xx means nothing, but I guess then we have found the least bad option.